File: frontend/src/types/recruiter.ts
// TypeScript interfaces for Recruiter Portal and Admin Panel

import type { Language } from "./coding";

export interface CompanyResponse {
  id: number;
  name: string;
  domain: string | null;
  logo_url: string | null;
  description: string | null;
  created_at: string;
}

export interface RecruiterResponse {
  id: number;
  user_id: number;
  company_id: number;
  title: string | null;
  created_at: string;
  company?: CompanyResponse;
}

export interface InterviewTemplateResponse {
  id: number;
  company_id: number;
  title: string;
  job_role: string;
  experience_level: string;
  duration_minutes: number;
  question_count: number;
  questions: Array<{ question: string }>;
  created_at: string;
}

export interface RecruiterInterviewResponse {
  id: number;
  recruiter_id: number;
  company_id: number;
  title: string;
  description: string | null;
  job_role: string;
  flow_config: Array<{ type: string; difficulty?: string }>;
  status: string;
  created_at: string;
}

export interface CandidateAssignmentResponse {
  id: number;
  user_id: number;
  recruiter_interview_id: number;
  status: "invited" | "in_progress" | "completed";
  invite_token: string | null;
  score: number | null;
  feedback_summary: string | null;
  assigned_at: string;
  completed_at: string | null;
  user_email: string | null;
  user_full_name: string | null;
  interview_title: string | null;
}

export interface UserAdminResponse {
  id: number;
  email: string;
  full_name: string | null;
  role: "candidate" | "recruiter" | "admin";
  is_active: boolean;
  created_at: string;
}

export interface AdminSettingsResponse {
  id: number;
  key: string;
  value: Record<string, unknown>;
  description: string | null;
  updated_at: string;
}

export interface AuditLogResponse {
  id: number;
  user_id: number | null;
  action: string;
  ip_address: string | null;
  details: Record<string, unknown>;
  created_at: string;
  user_email: string | null;
}

export interface PlatformAnalyticsResponse {
  total_candidates: number;
  total_recruiters: number;
  total_interviews_completed: number;
  average_score: number;
  provider_configurations: Record<string, string>;
  system_load: number;
}

// Aggregated Candidate Reports details
export interface CandidateMockInterviewReport {
  id: number;
  job_role: string;
  company: string | null;
  interview_type: string;
  difficulty: string;
  status: string;
  created_at: string;
  feedback: {
    overall_score: number;
    technical_score: number;
    communication_score: number;
    feedback_text: string;
  } | null;
}

export interface CandidateCodingSessionReport {
  id: number;
  question_title: string;
  language: Language;
  status: string;
  timer_seconds: number;
  updated_at: string;
}

export interface CandidateResumeReport {
  id: number;
  filename: string;
  ats_score: number;
  missing_skills: string[];
  uploaded_at: string;
}

export interface CandidateAggregateReports {
  interviews: CandidateMockInterviewReport[];
  coding: CandidateCodingSessionReport[];
  resumes: CandidateResumeReport[];
}

File: backend/app/schemas/recruiter.py
"""
Pydantic schemas for Recruiter Portal and Admin Panel.
"""

from datetime import datetime
from typing import Any

from pydantic import BaseModel, Field

# ---------------------------------------------------------------------------
# Company Schemas
# ---------------------------------------------------------------------------

class CompanyCreate(BaseModel):
    name: str = Field(..., max_length=255)
    domain: str | None = Field(None, max_length=255)
    logo_url: str | None = Field(None, max_length=500)
    description: str | None = None


class CompanyUpdate(BaseModel):
    name: str | None = Field(None, max_length=255)
    domain: str | None = Field(None, max_length=255)
    logo_url: str | None = Field(None, max_length=500)
    description: str | None = None


class CompanyResponse(BaseModel):
    id: int
    name: str
    domain: str | None = None
    logo_url: str | None = None
    description: str | None = None
    created_at: datetime

    class Config:
        from_attributes = True


# ---------------------------------------------------------------------------
# Recruiter Schemas
# ---------------------------------------------------------------------------

class RecruiterCreate(BaseModel):
    company_id: int
    title: str | None = Field(None, max_length=100)


class RecruiterResponse(BaseModel):
    id: int
    user_id: int
    company_id: int
    title: str | None = None
    created_at: datetime
    company: CompanyResponse | None = None

    class Config:
        from_attributes = True


# ---------------------------------------------------------------------------
# Interview Template Schemas
# ---------------------------------------------------------------------------

class InterviewTemplateCreate(BaseModel):
    title: str = Field(..., max_length=255)
    job_role: str = Field(..., max_length=100)
    experience_level: str = Field(..., max_length=50)
    duration_minutes: int = Field(30, ge=5, le=120)
    question_count: int = Field(5, ge=1, le=50)
    questions: list[dict[str, Any]] = []


class InterviewTemplateResponse(BaseModel):
    id: int
    company_id: int
    title: str
    job_role: str
    experience_level: str
    duration_minutes: int
    question_count: int
    questions: list[dict[str, Any]]
    created_at: datetime

    class Config:
        from_attributes = True


# ---------------------------------------------------------------------------
# Recruiter Interview Flow Schemas
# ---------------------------------------------------------------------------

class RecruiterInterviewCreate(BaseModel):
    title: str = Field(..., max_length=255)
    description: str | None = None
    job_role: str = Field(..., max_length=100)
    flow_config: list[dict[str, Any]] = []  # e.g., [{"type": "resume"}, {"type": "mock_interview"}, ...]


class RecruiterInterviewResponse(BaseModel):
    id: int
    recruiter_id: int
    company_id: int
    title: str
    description: str | None = None
    job_role: str
    flow_config: list[dict[str, Any]]
    status: str
    created_at: datetime

    class Config:
        from_attributes = True


# ---------------------------------------------------------------------------
# Candidate Assignment / Invites Schemas
# ---------------------------------------------------------------------------

class CandidateAssignmentCreate(BaseModel):
    email: str  # We invite via email (resolves to existing user or we mock invite)
    recruiter_interview_id: int


class CandidateAssignmentResponse(BaseModel):
    id: int
    user_id: int
    recruiter_interview_id: int
    status: str
    invite_token: str | None = None
    score: float | None = None
    feedback_summary: str | None = None
    assigned_at: datetime
    completed_at: datetime | None = None
    user_email: str | None = None
    user_full_name: str | None = None
    interview_title: str | None = None

    class Config:
        from_attributes = True


# ---------------------------------------------------------------------------
# Admin User Management Schemas
# ---------------------------------------------------------------------------

class UserAdminResponse(BaseModel):
    id: int
    email: str
    full_name: str | None = None
    role: str
    is_active: bool
    created_at: datetime

    class Config:
        from_attributes = True


class UserUpdateRole(BaseModel):
    role: str = Field(..., pattern=r"^(candidate|recruiter|admin)$")
    is_active: bool | None = None


# ---------------------------------------------------------------------------
# Admin Settings & Feature Flags Schemas
# ---------------------------------------------------------------------------

class AdminSettingsUpdate(BaseModel):
    value: dict[str, Any]
    description: str | None = None


class AdminSettingsResponse(BaseModel):
    id: int
    key: str
    value: dict[str, Any]
    description: str | None = None
    updated_at: datetime

    class Config:
        from_attributes = True


# ---------------------------------------------------------------------------
# Audit Log Schemas
# ---------------------------------------------------------------------------

class AuditLogResponse(BaseModel):
    id: int
    user_id: int | None = None
    action: str
    ip_address: str | None = None
    details: dict[str, Any]
    created_at: datetime
    user_email: str | None = None

    class Config:
        from_attributes = True


# ---------------------------------------------------------------------------
# Platform Analytics Schemas
# ---------------------------------------------------------------------------

class PlatformAnalyticsResponse(BaseModel):
    total_candidates: int
    total_recruiters: int
    total_interviews_completed: int
    average_score: float
    provider_configurations: dict[str, str]
    system_load: float
